### DISCUSSION TITLE
MSD7816 booting

### DISCUSSION THREAD

**Discussion by RDS5** [linux-chenxing/linux-chenxing.org#73] [General]
i have an old set top box and it has an MSD7816. looking up this chip, one of the only results seems to be someone [booting linux on here ](https://mstar.fandom.com/wiki/MSD7816)

there is uart on this device and i can access u-boot/mboot terminal, but all my attempts to boot linux on this device seem to have been unsuccessful. (i have been using this repository to build linux, just like they did, but it hasnt booted)

the device's SoC (MSD7816-L2) and board (HW-150PVR)

![20221123_185153_001_saved](https://user-images.githubusercontent.com/32397126/203667891-11884006-bd89-4030-a129-f1f6c5342d89.jpg)
![20221123_185528](https://user-images.githubusercontent.com/32397126/203667893-880c97bf-1c91-414d-884f-e40501a2d050.jpg)

the log when i try to boot linux (booting from usb, it gets stuck at "Starting kernel ...")

```
BOOTSPI[3456789AB3456789AB]_OK!decomp
_done
MDrv_PM_RtcGetCounter(CurrentCounter=0)

Hello U-Boot
board_init 


U-Boot 1.1.6 (Jan  7 2015 - 22:02:26)

Board: MSTAR KRNOUS (CPU Speed 552 MHz)
DRAM:  64 X 0 MBytes
U-Boot is running at DRAM 0x87610000
###############BOARD CONFIGURATION#####################
	DEFAULT ENBALE L2-Cache
	FPU(ENABLE)
   DDR_FREQUENCY(1066MHz)
###############BOARD CONFIGURATION#####################
Module: USB FAT FLASH SPI LOGO OSD ENV=SERIAL 
Flash is detected (0x0C00, 0xC8, 0x40, 0x16)
MDrv_SERFLASH_GetInfo()
	u32AccessWidth = 1
	u32TotalSize   = 4194304
	u32SecNum      = 64
	u32SecSize     = 65536
In:    serial
Out:   serial
Err:   serial
MDrv_IIC_InitPort0: OK
-----MDrv_Demod_Init start-----
IF FREQ 5000
INTERN_ATSCFW_VERSION:ED.0.9

 MDrv_Demod_Init 1 Tuner R820T Init sussec
MSVC00B000100100208768TH0000000T
MDrv_PNL_Init u32PnlRiuBaseAddr = BF200000
MDrv_PNL_Init u32PMRiuBaseAddr = BF000000
[XC,Version] 00423696
 DAC eTiming =9
HDMITx eTiming =13
HDMITx eTiming =13
u32ReadBuffVirAddr = A0000000, u32IntBuffVirAddr = A0100000, u32OutBuffVirAddr = A0730000
verJPD_SetStatus >>>>>>>>>>> w:720,  h:576,  p:720
GE_SetOnePixelMode
Hit any key to stop autoboot:  0 

<< MStar >># usb start
(Re)start USB...
Kronus 128-pin package
Check USB port[0]:
Host type:2 
scanning bus for devices... 1 USB Device(s) found
       scanning bus for storage devices... bulk max packet size:200
usb_stor_Bulk_max_lun: 0 
init_part begin
init_part end
1 Storage Device(s) found
<< MStar >># fatload usb 0 0x80100000 uimagelinuxbin
reading uimagelinuxbin

4088636 bytes read
<< MStar >># setenv bootargs console=ttyS0,115200
<< MStar >># bootm 0x80100000
## Booting image at 80100000 ...
   Image Name:   Linux
   Created:      2022-11-23  23:58:26 UTC
   Image Type:   MIPS Linux Kernel Image (lzo compressed)
   Data Size:    4088572 Bytes =  3.9 MB
   Load Address: 80100000
   Entry Point:  803A6FE0
   Verifying Checksum ... OK
   Uncompressing Kernel Image ... OK

Starting kernel ...


```

environment variables

```
<< MStar >># printenv
bootdelay=0
baudrate=115200
preboot=echo;echo Type "help" for more commands.
MS_BOARD=BD_MST124C_D01A_S
logo_cmd=boot_logo 0 0 1 1
AppMagicFlagAddr=0x80000
bootcmd=spi_rdc 0x80B00000 0x8001C 0x2AAC10; LzmaDec 0x80B00000 0x2AAC10 0x80000180 0x81000000; go 0x80000224;
info_exchange=spi
CUSTOMER_OUI=0x169B
AP_SW_MODEL=0x0001
AP_SW_VERSION=0x0001
HW_MODEL=0x0001
HW_VERSION=0x0001
panel_cmd=set_paneltype 18
USBUpgradeFlag=0
USBUpdateFail=1
USBUpdateFlag=0
stdin=serial
stdout=serial
stderr=serial

Environment size: 496/65532 bytes
```

**Comment by fifteenhex** [comment]
Hi,

That's pretty interesting as I've never tried to get the kernel here running on MIPS.
@kagaimiq has been adding lots of info about the MIPS chips so maybe they have added MIPS support? I couldn't find the fork if there is one... I'd love to add MIPS support to our branch and eventually get it mainlined though.

Cheers,

Daniel

**Reply by RDS5** [reply]
yes, they did boot linux on the MSD7816
they're the ones who made the MSD7816 page on fandom (ewogs30 made the edits, ewogs30 is a name @kagaimiq goes by) and a channel by the same name (ewogs30) uploaded a video of them booting debian on a titania4 board. on top of that, they mentioned that they had tried using the [FPU on MSD7816 on 5.10.15](https://linux-chenxing.org/kronus/)

**Reply by kagaimiq** [reply]
Yep, that's actually true.

Regarding the kernel, i currently only have a local clone of the `mstar_v6_0_rebase` branch of the linux-chenxing kernel.

And yes, i was the one who made that page (as well as few other edits) on fandom (- months before i got to know that linux-chenxing was ever a thing other than some patches sent into the Linux kernel mailing lists)

**Reply by RDS5** [reply]
wait but, how did you make the kernel boot on this chip?
i cant for the life of me figure it out (im probably making a really obvious mistake that i just cant understand)

**Reply by kagaimiq** [reply]
You know, it isn't something impossible to do.

Making Linux run on that chip was just a matter of adding a new MIPS platform, which is basically just adding some entries in the Kconfig/Kbuild/Makefile, making up the correct device tree, and that's it.

Then the kernel is configured to use the new MIPS platform, and then it's compiled.

When it's done compiling, you make a bootable image out of the vmlinux, copy it on your USB flash, and then you just enter in U-Boot the commands to boot that image and it boots.

That's it.

**Reply by fifteenhex** [reply]
Hi kagaimiq,

Would you mind sharing your patches somewhere? I would like to add them to my branch if possible.
Then I can get one of these MIPS chips and test that the drivers we have work there too.

**Reply by kagaimiq** [reply]
Here are my patches: [msmipsx.tar.gz](https://github.com/linux-chenxing/linux-chenxing.org/files/10097080/msmipsx.tar.gz)
It's just a very basic and hacky support for kronus and titania4, that's it.

**Reply by maniekx86** [reply]
I have a similar board with MSD7818. I compiled Linux kernel with @kagaimiq patches (in menuconfig selected board type to MStar soc, enabled kronus support, MSD7816 test dtb and enabled MStar interrupt controller) and tried booting - stuck as original poster. I am sure that i am doing something wrong (wrong kernel config or repo used/wrong load address)? 
```
<< MStar >># fatload usb 0 0x81000000 uImage.bin
reading uImage.bin

3990224 bytes read
<< MStar >># bootm 0x81000000 
## Booting image at 81000000 ...
   Image Name:   Linux-6.0.0-dirty
   Created:      2022-12-01  18:16:21 UTC
   Image Type:   MIPS Linux Kernel Image (uncompressed)
   Data Size:    3990160 Bytes =  3.8 MB
   Load Address: 80000000
   Entry Point:  802B808C
   Verifying Checksum ... OK
OK

Starting kernel ...

```
Also note - I could not compile these patches with linux-chenxing/linux repo as it throws compilation errors:
```
drivers/clk/mstar/clk-msc313-mux.c: In function ‘msc313_mux_mux_get_parent’:
drivers/clk/mstar/clk-msc313-mux.c:27:2: error: implicit declaration of function ‘regmap_field_read’; did you mean ‘regmap_field_write’? [-Werror=implicit-function-declaration]
   27 |  regmap_field_read(mux->mux, &index);
      |  ^~~~~~~~~~~~~~~~~
      |  regmap_field_write
drivers/clk/mstar/clk-msc313-mux.c: In function ‘msc313_mux_register_muxes’:
drivers/clk/mstar/clk-msc313-mux.c:163:16: error: variable ‘gate_field’ has initializer but incomplete type
  163 |   const struct reg_field gate_field = REG_FIELD(mux_data->offset,
      |                ^~~~~~~~~
drivers/clk/mstar/clk-msc313-mux.c:163:39: error: implicit declaration of function ‘REG_FIELD’ [-Werror=implicit-function-declaration]
  163 |   const struct reg_field gate_field = REG_FIELD(mux_data->offset,
      |                                       ^~~~~~~~~
drivers/clk/mstar/clk-msc313-mux.c:163:26: error: storage size of ‘gate_field’ isn’t known
  163 |   const struct reg_field gate_field = REG_FIELD(mux_data->offset,
      |                          ^~~~~~~~~~
drivers/clk/mstar/clk-msc313-mux.c:165:16: error: variable ‘mux_field’ has initializer but incomplete type
  165 |   const struct reg_field mux_field = REG_FIELD(mux_data->offset,
      |                ^~~~~~~~~
drivers/clk/mstar/clk-msc313-mux.c:165:26: error: storage size of ‘mux_field’ isn’t known
  165 |   const struct reg_field mux_field = REG_FIELD(mux_data->offset,
      |                          ^~~~~~~~~
drivers/clk/mstar/clk-msc313-mux.c:167:16: error: variable ‘deglitch_field’ has initializer but incomplete type
  167 |   const struct reg_field deglitch_field = REG_FIELD(mux_data->offset,
      |                ^~~~~~~~~
drivers/clk/mstar/clk-msc313-mux.c:167:26: error: storage size of ‘deglitch_field’ isn’t known
  167 |   const struct reg_field deglitch_field = REG_FIELD(mux_data->offset,
      |                          ^~~~~~~~~~~~~~
drivers/clk/mstar/clk-msc313-mux.c:174:16: error: implicit declaration of function ‘devm_regmap_field_alloc’ [-Werror=implicit-function-declaration]
  174 |    mux->gate = devm_regmap_field_alloc(dev, regmap, gate_field);
      |                ^~~~~~~~~~~~~~~~~~~~~~~
drivers/clk/mstar/clk-msc313-mux.c:167:26: warning: unused variable ‘deglitch_field’ [-Wunused-variable]
  167 |   const struct reg_field deglitch_field = REG_FIELD(mux_data->offset,
      |                          ^~~~~~~~~~~~~~
drivers/clk/mstar/clk-msc313-mux.c:165:26: warning: unused variable ‘mux_field’ [-Wunused-variable]
  165 |   const struct reg_field mux_field = REG_FIELD(mux_data->offset,
      |                          ^~~~~~~~~
drivers/clk/mstar/clk-msc313-mux.c:163:26: warning: unused variable ‘gate_field’ [-Wunused-variable]
  163 |   const struct reg_field gate_field = REG_FIELD(mux_data->offset,
      |                          ^~~~~~~~~~
  CC      drivers/tty/vt/vt_ioctl.o
cc1: some warnings being treated as errors
```

**Reply by kagaimiq** [reply]
First of all, yes, you need to use the linux-chenxing kernel, as there's much stuff that depends on this fork after all. these patches i put there were for the `mstar_v6_1_rebase` branch.

To resolve that build issue you need to enable `CONFIG_MFD_SYSCON` so that it will enable regmap thing, and it should compile fine now.

[here](https://github.com/linux-chenxing/linux-chenxing.org/files/10095221/msmips.tar.gz) i have a config that works, although it was made for an older patch variant, so be sure to reselect the "msd7816 test board" once again, and maybe other changes as well.

Also you'd better to disable the video initialization (by doing `setenv logo_cmd nop; saveenv; reset`); or otherwise linux will eventually get to the portion of memory where the VE stores its framebuffer (VE is the thing that outputs composite video) -- and so any data that it will try to store there will be corrupted.

Or move the `framebuffer` and `fb-memory` node's addresses to `0x2100000` and add a node into `reserved-memory` with `reg = <0x3200000 0x1a0000>;` (which reserves the memory region where VE lives), and you will have sorta working framebuffer too (but with wrong pixel format - GOP expects YUV422 while Linux expects RGB565... i really need to figure out that video subsystem mess and make a proper DRM driver)

**Reply by maniekx86** [reply]
Thanks! Got it working by using your tips and config. It's really cool to see Linux running on this chip. I loaded uImage into 0x81000000 and booted it with bootm. 

Note about framebuffer in my case: on video output I see half top of screen actually being changed by /dev/fb0 contents (yet still wrong format) and bottom probably some RAM as it is garbage and changes when running some programs. I am really interested seeing in future video output working.

Also a question: As I have own blog could I in future post article and guide about running Linux on this chip using your configs?

**Reply by fifteenhex** [reply]
> Also a question: As I have own blog could I in future post article and guide about running Linux on this chip using your configs?

I have a request if you have time.. can you put the working tree you have on github? It'd be really nice to eventually send the MIPS patches to mainline and get a bit more attention. We have drivers for almost all of the hardware in these chips now but not enough people to get it all cleaned up and in mainline.

**Reply by maniekx86** [reply]
Yeah, I can. I am not sure have I done it right as you wanted, but I forked repo and applied patches to it (https://github.com/maniekx86/linux/tree/mstar_v6_1_rebase), also fixed half screen framebuffer issue by changing dts file.

**Reply by maniekx86** [reply]
Further playing with this chip: I managed to autoload Linux on boot via bootcmd - it's disables WDT (as it was reseting CPU when using bootcmd) using mw (memory write) commands (this was kinda tricky as u-boot here is without wdt command, but managed to do this using references from MStar source codes), loading image from USB to mem and booting it.

**Reply by fifteenhex** [reply]
Hi,

We have an WDT driver and it'll probably work for the MIPS chips too. Maybe you could have a go at wiring it up?

**Reply by maniekx86** [reply]
I could not get WDT driver working probably because I don't have enough experience with kernel sources and device drivers. For now I am experimenting with board, but yet it's still cool to experiment and test stuff.

**Comment by fifteenhex** [comment]
I've pulled two of the MStar MIPS patches into the linux-chenxing tree.. I need to check the intc one a bit more.

Anyhow.. It'd be really cool if someone wanted to step up to send the patches to mainline and become the maintainer for MStar MIPS.

**Comment by mztulip** [comment]
Hello, I booted sucessfully kernel 6.7. Do you know if the is any driver for i2c?
Digging through device tree I found something like this:
`	miic1: miic@206840 {
		compatible = "mstar,msd7816-miic";
		reg = <0x206840 0x40>;
		interrupts-extended = <&intc_irq 1 IRQ_TYPE_LEVEL_HIGH>;
		clocks = <&mpll_24m>; // TODO, clkgen2!
	};`
Driver mstar,msd7816-miic is not existing in source code.

**Comment by cybersnow02** [comment]
Hello,
I have a Homeworx HW-150PVR V2 from 2014. The Mediasonic team sent the wrong firmware for this model and it bricked the unit. The company refuse to send the old firmware. First, is it possible to reflash a bricked unit without a programmer? Can I just put a firmware file on a USB key and reupload it by itself at init boot? Do you guys have an old firmware for this unit? thanks

**Reply by mztulip** [reply]
In my model, everything can be done using a USB drive and U-Boot commands. You can both dump and flash the firmware, although it is not a one-command process. I can upload the firmware from my device, but my model is different from yours.(I have Wiwa hd95)

**Reply by cybersnow02** [reply]
Well, someone found for me the right firmware for my unit. But, it seems to be hardbricked. When I plug the USB key and powerOn the unit, the USB key light rapidly turn off. It look like the MBoot/Uboot is freezing at first loading. I cannot find a memory programmer to reupload the right firmware. Is there anything I can do with it or can I just put it in the E-Bin?
